File: frontend-web/src/app/games/xiangqi/board/xiangqi-board.ts
import {
  ChangeDetectionStrategy,
  Component,
  computed,
  effect,
  input,
  output,
  signal,
} from '@angular/core';
import { TranslocoPipe } from '@jsverse/transloco';
import type { MoveDto, RoomState } from '../../../core/api/models/room.model';
import {
  BLACK,
  pieceAt,
  positionAfter,
  RED,
  XIANGQI_COLS,
  XIANGQI_ROWS,
  type XiangqiPiece,
  type XiangqiPieceType,
  type XiangqiSide,
} from '../position';

export interface BoardPoint {
  readonly row: number;
  readonly col: number;
}

export interface PieceMoveEvent {
  readonly from: BoardPoint;
  readonly to: BoardPoint;
}

/**
 * Piece glyphs.
 *
 * These are **graphics, not copy** — they are what a 象棋 piece looks like in every
 * locale, the way a chess knight is a horse's head everywhere. They therefore stay
 * here rather than in the locale files. The *spoken* name of each piece is a
 * different thing and does go through i18n (see `pieceNameKey`).
 */
const GLYPHS: Record<XiangqiSide, Record<XiangqiPieceType, string>> = {
  // Stone.Black is 红 — see position.ts and design D3.
  Black: {
    general: '帥',
    advisor: '仕',
    elephant: '相',
    horse: '傌',
    chariot: '俥',
    cannon: '炮',
    soldier: '兵',
  },
  White: {
    general: '將',
    advisor: '士',
    elephant: '象',
    horse: '馬',
    chariot: '車',
    cannon: '砲',
    soldier: '卒',
  },
};

/**
 * Grid lines, in a coordinate space where intersection (row, col) sits at (col, row).
 *
 * Built once as a compound path: ten ranks, two full outer files, seven files broken
 * by the river, and the two palaces' diagonals. The 兵/炮 corner brackets a printed
 * board carries are deliberately left out — they are decoration, and drawing them
 * would triple this path for no information.
 */
const GRID_PATH = (() => {
  const parts: string[] = [];
  for (let r = 0; r < XIANGQI_ROWS; r++) parts.push(`M0 ${r}H${XIANGQI_COLS - 1}`);
  parts.push(`M0 0V${XIANGQI_ROWS - 1}`, `M${XIANGQI_COLS - 1} 0V${XIANGQI_ROWS - 1}`);
  for (let c = 1; c < XIANGQI_COLS - 1; c++) parts.push(`M${c} 0V4`, `M${c} 5V${XIANGQI_ROWS - 1}`);
  parts.push('M3 0L5 2', 'M5 0L3 2', 'M3 7L5 9', 'M5 7L3 9');
  return parts.join(' ');
})();

/**
 * 中国象棋 board — 10×9 intersections, two-step piece movement.
 *
 * **Not a parameterisation of `Board`.** `Board` renders "is there a stone in this
 * cell" and moves in one click; this renders "which piece is on this intersection"
 * and moves in two. Folding them together would produce a presentational component
 * with a `gameKey` branch inside it, which is the shape this game exists to avoid.
 *
 * It judges **no** move legality (design D2). It does exactly two things that need
 * no rules: you can only pick up your own piece, and the board is read-only when it
 * is not your turn. Everything else is the server's call — an illegal target comes
 * back as a `HubException` and the container shows a toast. The alternative was a
 * TypeScript port of the rules, i.e. a second source of truth whose divergence would
 * read to the player as a bug and which nothing here would ever detect.
 *
 * Purely presentational: no `GameHubService`, no `GameCatalogService`, no router.
 */
@Component({
  selector: 'app-xiangqi-board',
  standalone: true,
  imports: [TranslocoPipe],
  templateUrl: './xiangqi-board.html',
  styles: [':host { display: block; width: 100%; }'],
  // Escape lives on the host rather than on the board div: keydown bubbles out of
  // whichever intersection has focus, and the host is not an interactive element
  // pretending to be focusable.
  host: { '(keydown.escape)': 'clearSelection()' },
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class XiangqiBoard {
  readonly state = input<RoomState | null>(null);
  readonly mySide = input<'black' | 'white' | 'spectator'>('spectator');
  readonly submitting = input<boolean>(false);
  readonly = input<boolean>(false);
  readonly pieceMove = output<PieceMoveEvent>();

  protected readonly gridPath = GRID_PATH;
  protected readonly rowIndices = Array.from({ length: XIANGQI_ROWS }, (_, i) => i);
  protected readonly colIndices = Array.from({ length: XIANGQI_COLS }, (_, i) => i);

  /** The origin the player has picked up, if any. Cleared whenever a ply lands. */
  private readonly selected = signal<BoardPoint | null>(null);
  private previousMoveCount = -1;

  private readonly moves = computed<readonly MoveDto[]>(() => this.state()?.game?.moves ?? []);
  protected readonly position = computed(() => positionAfter(this.moves()));

  protected readonly lastMove = computed<MoveDto | null>(() => {
    const moves = this.moves();
    return moves.length > 0 ? moves[moves.length - 1] : null;
  });

  /** `Stone` value of the seat the viewer occupies; `null` for spectators. */
  private readonly myStone = computed<XiangqiSide | null>(() => {
    const side = this.mySide();
    if (side === 'black') return RED;
    if (side === 'white') return BLACK;
    return null;
  });

  private readonly myTurn = computed<boolean>(() => {
    const mine = this.myStone();
    return mine !== null && this.state()?.game?.currentTurn === mine;
  });

  protected readonly boardDisabled = computed<boolean>(
    () =>
      this.readonly() ||
      this.submitting() ||
      this.mySide() === 'spectator' ||
      this.state()?.status !== 'Playing' ||
      !this.myTurn(),
  );

  constructor() {
    // A landed ply — mine or the opponent's — makes the held origin meaningless.
    // Rejections do NOT land a ply, so a refused move keeps the piece in hand:
    // the player almost always wants a different target, not to hunt for the piece again.
    effect(() => {
      const n = this.moves().length;
      if (this.previousMoveCount !== -1 && n !== this.previousMoveCount) this.selected.set(null);
      this.previousMoveCount = n;
    });
  }

  protected pieceAtPoint(row: number, col: number): XiangqiPiece | null {
    return pieceAt(this.position(), row, col);
  }

  protected glyph(row: number, col: number): string {
    const piece = this.pieceAtPoint(row, col);
    return piece ? GLYPHS[piece.side][piece.type] : '';
  }

  protected isRed(row: number, col: number): boolean {
    return this.pieceAtPoint(row, col)?.side === RED;
  }

  protected isSelected(row: number, col: number): boolean {
    const sel = this.selected();
    return sel !== null && sel.row === row && sel.col === col;
  }

  protected isLastTo(row: number, col: number): boolean {
    const last = this.lastMove();
    return last !== null && last.row === row && last.col === col;
  }

  protected isLastFrom(row: number, col: number): boolean {
    const last = this.lastMove();
    return last != null && last.fromRow === row && last.fromCol === col;
  }

  /** Translation key for the spoken name of whatever is on this intersection. */
  protected pieceNameKey(row: number, col: number): string {
    const piece = this.pieceAtPoint(row, col);
    if (!piece) return 'xiangqi.board.empty';
    return `xiangqi.piece.${piece.side === RED ? 'red' : 'black'}-${piece.type}`;
  }

  /**
   * A point is reachable only when clicking it would do something: it holds one of
   * my pieces (pick up / re-pick), or a piece is already in hand (move there).
   */
  protected pointDisabled(row: number, col: number): boolean {
    if (this.boardDisabled()) return true;
    if (this.selected() !== null) return false;
    return this.pieceAtPoint(row, col)?.side !== this.myStone();
  }

  protected handleClick(row: number, col: number): void {
    if (this.pointDisabled(row, col)) return;

    if (this.isSelected(row, col)) {
      this.selected.set(null);
      return;
    }

    if (this.pieceAtPoint(row, col)?.side === this.myStone()) {
      // Another of my own pieces: re-pick it. Emitting here would be a request the
      // server is certain to refuse, and "capture your own piece" is not a thing.
      this.selected.set({ row, col });
      return;
    }

    const from = this.selected();
    if (from === null) return;
    this.pieceMove.emit({ from, to: { row, col } });
  }

  protected clearSelection(): void {
    this.selected.set(null);
  }
}

File: backend/src/Gewu.Domain/Games/Xiangqi/XiangqiBoard.cs
using Gewu.Domain.Enums;
using Gewu.Domain.ValueObjects;

namespace Gewu.Domain.Games.Xiangqi;

/// <summary>
/// 中国象棋盘面,10 行 × 9 列。**规则内部使用** —— 聚合根看不到它,这正是
/// <c>generalize-match-domain</c> 把盘面语义下沉进 <c>IGameRules</c> 的目的。
/// <para>
/// 方位约定(见 <see cref="XiangqiRules"/>):<see cref="Stone.Black"/> 是**红方**,占第 5–9 行
/// (第 9 行是它的底线);<see cref="Stone.White"/> 是黑方,占第 0–4 行。楚河汉界在第 4 与第 5 行之间。
/// </para>
/// <para>
/// 非线程安全 —— 每次 <c>Apply</c> 从历史新建一块,不跨调用共享。
/// </para>
/// </summary>
internal sealed class XiangqiBoard
{
    /// <summary>行数。</summary>
    internal const int RowCount = 10;

    /// <summary>列数。</summary>
    internal const int ColCount = 9;

    private readonly XiangqiPiece?[] _cells = new XiangqiPiece?[RowCount * ColCount];

    private XiangqiBoard()
    {
    }

    /// <summary>摆好开局的棋盘。</summary>
    /// <summary>一块空盘 —— 残局从设置摆子时的起点。</summary>
    /// <returns>没有任何棋子的局面。</returns>
    internal static XiangqiBoard Empty() => new();

    internal static XiangqiBoard Initial()
    {
        var board = new XiangqiBoard();

        // 黑方(Stone.White)在上,第 0 行是它的底线。
        board.PlaceBackRank(0, Stone.White);
        board.Set(2, 1, new XiangqiPiece(XiangqiPieceType.Cannon, Stone.White));
        board.Set(2, 7, new XiangqiPiece(XiangqiPieceType.Cannon, Stone.White));
        for (var col = 0; col < ColCount; col += 2)
        {
            board.Set(3, col, new XiangqiPiece(XiangqiPieceType.Soldier, Stone.White));
        }

        // 红方(Stone.Black)在下,第 9 行是它的底线。红先手 —— 与 Game.CurrentTurn 的初值对齐。
        board.PlaceBackRank(9, Stone.Black);
        board.Set(7, 1, new XiangqiPiece(XiangqiPieceType.Cannon, Stone.Black));
        board.Set(7, 7, new XiangqiPiece(XiangqiPieceType.Cannon, Stone.Black));
        for (var col = 0; col < ColCount; col += 2)
        {
            board.Set(6, col, new XiangqiPiece(XiangqiPieceType.Soldier, Stone.Black));
        }

        return board;
    }

    private void PlaceBackRank(int row, Stone side)
    {
        XiangqiPieceType[] order =
        [
            XiangqiPieceType.Chariot, XiangqiPieceType.Horse, XiangqiPieceType.Elephant,
            XiangqiPieceType.Advisor, XiangqiPieceType.General, XiangqiPieceType.Advisor,
            XiangqiPieceType.Elephant, XiangqiPieceType.Horse, XiangqiPieceType.Chariot,
        ];
        for (var col = 0; col < ColCount; col++)
        {
            Set(row, col, new XiangqiPiece(order[col], side));
        }
    }

    /// <summary>该坐标是否在盘内。</summary>
    internal static bool InBounds(Position p) => p.Row < RowCount && p.Col < ColCount;

    /// <summary>取某格的棋子;空格返回 <c>null</c>。</summary>
    internal XiangqiPiece? At(Position p) => _cells[(p.Row * ColCount) + p.Col];

    /// <summary>取某格的棋子;空格返回 <c>null</c>。</summary>
    internal XiangqiPiece? At(int row, int col) => _cells[(row * ColCount) + col];

    internal void Set(int row, int col, XiangqiPiece? piece) => _cells[(row * ColCount) + col] = piece;

    /// <summary>
    /// 走一步 —— **不做任何校验**,合法性由 <see cref="XiangqiRules"/> 判定。
    /// 起点清空,终点覆盖(被吃的子就此消失)。
    /// </summary>
    internal void Move(Position from, Position to)
    {
        var piece = At(from);
        Set(from.Row, from.Col, null);
        Set(to.Row, to.Col, piece);
    }

    /// <summary>复制一份,供「试走一步看会不会自将」使用。</summary>
    internal XiangqiBoard Clone()
    {
        var clone = new XiangqiBoard();
        Array.Copy(_cells, clone._cells, _cells.Length);
        return clone;
    }

    /// <summary>
    /// 两块盘面上的子完全相同吗 —— 「同一个局面出现过几次」要数的就是它。
    /// <para>
    /// 逐格比 90 格,而不是先算一个指纹字符串:指纹要么有碰撞(那时计数会**多**数,
    /// 而多数的表现是一步合法的棋被拒),要么就是这 90 格本身。而每走一步都为整段历史
    /// 生成 90 字符的串是真实的垃圾,只为省下一个不慢的循环。
    /// </para>
    /// <para>
    /// <b>它不判「轮到谁」——那不在盘面里。</b> 一般而言同一块盘面在红走完与黑走完之后是两个
    /// 不同的局面;而长将上限那一处不需要区分,理由在
    /// <c>XiangqiRules.CountEarlierOccurrences</c> 上 —— 一个「对手被将」的盘面只可能由本方
    /// 走出来。别处要区分的话,MUST 自己把那件事算进去。
    /// </para>
    /// </summary>
    /// <param name="other">另一块盘面。</param>
    /// <returns>每一格都相同则为 <c>true</c>。</returns>
    internal bool SamePosition(XiangqiBoard other)
    {
        for (var i = 0; i < _cells.Length; i++)
        {
            if (_cells[i] != other._cells[i])
            {
                return false;
            }
        }
        return true;
    }

    /// <summary>找某方的将帅;被吃掉(理论上不该发生)时返回 <c>null</c>。</summary>
    internal Position? FindGeneral(Stone side)
    {
        for (var row = 0; row < RowCount; row++)
        {
            for (var col = 0; col < ColCount; col++)
            {
                if (At(row, col) is { Type: XiangqiPieceType.General } g && g.Side == side)
                {
                    return new Position(row, col);
                }
            }
        }
        return null;
    }

    /// <summary>枚举某方所有棋子的位置。</summary>
    internal IEnumerable<(Position Position, XiangqiPiece Piece)> PiecesOf(Stone side)
    {
        for (var row = 0; row < RowCount; row++)
        {
            for (var col = 0; col < ColCount; col++)
            {
                if (At(row, col) is { } piece && piece.Side == side)
                {
                    yield return (new Position(row, col), piece);
                }
            }
        }
    }

    /// <summary>两格之间(同行或同列,不含两端)夹着几个子。不同行不同列时返回 <c>-1</c>。</summary>
    internal int CountBetween(Position a, Position b)
    {
        if (a.Row != b.Row && a.Col != b.Col)
        {
            return -1;
        }

        var stepRow = Math.Sign(b.Row - a.Row);
        var stepCol = Math.Sign(b.Col - a.Col);
        var count = 0;

        var row = a.Row + stepRow;
        var col = a.Col + stepCol;
        while (row != b.Row || col != b.Col)
        {
            if (At(row, col) is not null)
            {
                count++;
            }
            row += stepRow;
            col += stepCol;
        }
        return count;
    }
}
